### PR TITLE
Speed-up compilation.

### PR DESCRIPTION
**Pull request by ryzhyk**
See commits for details.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Changelog
+All notable changes to this project will be documented in this file.
+
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
+
+## [Unreleased]
```
Comment (mbudiu-vmw) [inline CHANGELOG.md:6]: Maybe you want to add a few entries for previous releases; at least the first release should be here.
What is OOM?

File: app/Main.hs [hunk 2]
```diff
@@ -155,13 +158,26 @@ main = do
...
-            when (confRunRustfmt config') $
-                runCommandReportingErr "rustfmt" "cargo" ["fmt", "--all"] $ Just (confOutputDir config')
+            timeAction ("compiling " ++ show (confDatalogFile config)) $ do
+                compileProg config'
+                -- Run rustfmt on the generated code if it's enabled
+                when (confRunRustfmt config') $
+                    runCommandReportingErr "rustfmt" "cargo" ["fmt", "--all"] $ Just (confOutputDir config')
+
+-- Perform IO action and measure its duration.
+timeAction :: String -> IO () -> IO ()
```
Comment (mbudiu-vmw) [inline app/Main.hs:172]: Make the output conditional on some option (-verbose?).

File: src/Language/DifferentialDatalog/Expr.hs [hunk 3]
```diff
@@ -586,3 +589,144 @@ exprInjectStringConversion d e t = do
...
+    m' <- exprFoldWithLocatorM f (ELocator $ 0:path) m
+    cs' <- mapIdxM (\(e1, e2) i -> (,) <$> (exprFoldWithLocatorM f (ELocator $ (1+2*i):path) e1) <*>
+                                           (exprFoldWithLocatorM f (ELocator $ (2+2*i):path) e2)) cs
+    f loc $ EMatch p m' cs'
+exprFoldWithLocatorM' f loc                   (EVarDecl p v)          = f loc $ EVarDecl p v
+exprFoldWithLocatorM' f loc@(ELocator path) (ESeq p l r)            =
+    f loc =<< ESeq p <$> exprFoldWithLocatorM f (ELocator $ 0:path) l <*>
+                         exprFoldWithLocatorM f (ELocator $ 1:path) r
+exprFoldWithLocatorM' f loc@(ELocator path) (EITE p i t el)         =
+    -- 0: if-confition
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/Expr.hs:642]: typo in comment

File: src/Language/DifferentialDatalog/NS.hs [hunk 4]
```diff
@@ -150,7 +151,9 @@ getRelation :: DatalogProgram -> String -> Relation
 getRelation d n = fromJust $ lookupRelation d n
 
 arg2v :: Function -> FuncArg -> Var
-arg2v f a = ArgVar f (name a)
+arg2v f a = ArgVar f i (name a)
+    where
+    i = fromJust $ findIndex ((== name a) . name) $ funcArgs f
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/NS.hs:156]: these linear scans are something that you should perhaps keep an eye for.

File: src/Language/DifferentialDatalog/TypeInference.hs [hunk 5]
```diff
@@ -866,9 +895,9 @@ exprConstraints_ de@(DDExpr ctx (E e@EBinOp{..})) | elem exprBOp [Eq, Neq, Lt, L
                                                 | elem exprBOp [ShiftL, ShiftR] = do
     addConstraint =<< deIsInt l
     -- If the type of 'r' cannot be inferred, default to 'u32'
-    isbits <- deIsBits r
-    def <- tvarTypeOfExpr r <==== TEBit (IConst 32)
-    addConstraint $ isbits {cDefault = Just [def]}
+    --isbits <- deIsBits r
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/TypeInference.hs:898]: delete?

File: src/Language/DifferentialDatalog/Unification.hs [hunk 6]
```diff
@@ -50,11 +50,10 @@ module Language.DifferentialDatalog.Unification(
 where
 
 import Control.Monad.Except
-import Control.Monad.State
 import Data.List
 import qualified Data.Map as M
 import Data.Maybe
---import Debug.Trace
+-- import Debug.Trace
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/Unification.hs:56]: isn't it interesting that this program is both shorter and faster?

File: test/datalog_tests/function.fail.ast.expected [hunk 7]
```diff
@@ -111,9 +111,16 @@ Only non-extern functions can be overloaded.
...
 ^
 
-error: ./test/datalog_tests/function.fail.dl:8:21-8:39: expression '(ddlog_std::nth(components, 1))' must be of type 'Result<>', but its type is 'ddlog_std::Option<string>'
-    var last_name = components.nth(1)?;
-                    ^^^^^^^^^^^^^^^^^^
+error: ./test/datalog_tests/function.fail.dl:6:5-10:1: Type mismatch:
+expected type: ddlog_std::Result<(string,string),string>
+actual type: ddlog_std::Option<('a17,'a28)>
+in
+expression '{(var components = (/*ddlog_std::split, internment::split*/(name, " ")));
```
Comment (mbudiu-vmw) [inline test/datalog_tests/function.fail.ast.expected:118]: I don't understand where this embedded comment comes from.
Are you showing a list of candidate functions that match?
The comment syntax is not the most appropriate.
Perhaps something like `One of [ddlog_std::split, internment::split]`

**Review by mbudiu-vmw — APPROVED** [review]

File: app/Main.hs (hunk 2 above)
Comment (ryzhyk) [reply app/Main.hs:172]: What's wrong with always printing it?  Rust compiler does this bu default, and I find it useful, also as explicit indication that compilation succeeded.

Comment (mbudiu-vmw) [reply app/Main.hs:172]: The Unix philosophy is to print nothing on success.
If you write tools that consume the output of the compiler you may expect that there is no output if there are no errors or warnings. But this is certainly a low priority item; we can wait until someone asks for a -quiet flag.

File: CHANGELOG.md (hunk 1 above)
Comment (ryzhyk) [reply CHANGELOG.md:6]: order-of-magnitude. I'll convert this into a release entry when creating a release.

**Pull request merged**